File: container-with-most-water/container-with-most-water.rb
# @param {Integer[]} height
# @return {Integer}
def max_area(height)
  result = 0
  i = 0
  j = height.size - 1

  while i < j do
    min_possible = [height[i], height[j]].min
    water_i_j = (j - i) * min_possible

    result = [result, water_i_j].max

    if height[i] < height[j]
      i += 1
    else
      j -= 1
    end
  end

  result
end

n = gets.to_i
height = gets.chomp.split.map(&:to_i)

puts max_area(height)

File: container-with-most-water/container-with-most-water.cpp
#include <iostream>
#include <vector>

#define MAX_HEIGHT 100000

using namespace std;

class Solution {
public:
    int maxArea(vector<int>& height) {
      short int min_possible;
      int water_i_j, result = 0;

      for(int i = 0, j = height.size() - 1; i < j;){
        min_possible = min(height[i], height[j]);
        water_i_j = (j - i) * min_possible;

        result = max(result, water_i_j);

        height[i] < height[j] ? i++ : j--;
      }

      return result;
    }
};

int main(){
  int i, n, max;
  vector<int> height(MAX_HEIGHT);
  Solution solution;

  cin >> n;

  for(i = 0; i < n; i++)
    cin >> height[i];

  max = solution.maxArea(height);
  cout << max;

  return 0;
}
